src/server/ua_subscription_events.c

```diff
@@ -409,6 +409,593 @@ orOperator(UA_Server *server, UA_Session *session,
+        {2, 2, -1, -1, 1, -1, 1, -1, 0, 1, 1, -1, 2, -1, 2, -1, -1, 2, 1, 1, -1},
+        {2, 2, -1, -1, 1, -1, 1, -1, 2, 0, 1, -1, 2, 2, 2, -1, -1, 2, 2, 1, -1},
+        {2, 2, -1, -1, 1, -1, 1, -1, 2, 2, 0, -1, 2, 2, 2, -1, -1, 2, 2, 2, -1},
+        {-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, 0, -1, -1, 1, -1, -1, -1, -1, -1, -1},
+        {2, 2, -1, -1, 1, -1, 1, -1, 1, 1, 1, -1, 0, -1, 2, -1, -1, 1, 1, 1, -1},
+        {-1, -1, -1, -1, -1, -1, -1, -1, -1, 1, 1, -1, -1, 0, -1, -1, -1, 2, 1, 1, -1},
+        {1, 1, -1, 2, 1, 2, 1, 1, 1, 1, 1, 2, 1, -1, 0, 2, 2, 1, 1, 1, -1},
+        {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 1, 0, -1, -1, -1, -1, -1},
+        {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 1, 1, 0, -1, -1, -1, -1},
+        {2, 2, -1, -1, 1, -1, 1, -1, 1, 1, 1, -1, 2, 1, 2, -1, -1, 0, 1, 1, -1},
+        {2, 2, -1, -1, 1, -1, 1, -1, 2, 1, 1, -1, 2, 2, 2, -1, -1, 2, 0, 1, -1},
+        {2, 2, -1, -1, 1, -1, 1, -1, 2, 2, 1, -1, 2, 2, 2, -1, -1, 2, 2, 0, -1},
+        {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 0}
+    };
+
+    if(firstOperatorTypeKindIndex == UA_BYTE_MAX || secondOperatorTypeKindIndex == UA_BYTE_MAX)
+        return -1;
+
+    return convertLookup[firstOperatorTypeKindIndex][secondOperatorTypeKindIndex];
+}
+
+/* Compare operation for equal, gt, le, gte, lee
+ * UA_STATUSCODE_GOOD if the comparison was true
+ * UA_STATUSCODE_BADNOMATCH if the comparison was false
+ * UA_STATUSCODE_BADFILTEROPERATORINVALID for invalid operators
+ * UA_STATUSCODE_BADTYPEMISMATCH if one of the operands was not numeric
+ * ToDo Array-Casting
+ */
+static UA_StatusCode
+compareOperation(UA_Variant *firstOperand, UA_Variant *secondOperand, UA_FilterOperator op) {
+    //get precedence of the operand types
+    UA_Int16 firstOperand_precedence = UA_DataType_getPrecedence(firstOperand->type);
+    UA_Int16 secondOperand_precedence = UA_DataType_getPrecedence(secondOperand->type);
+    //if the types are not equal and one of the precedence-ranks is -1, then there is
+    //no implicit conversion possible and therefore no compare
+    if(!UA_NodeId_equal(&firstOperand->type->typeId, &secondOperand->type->typeId) &&
+       (firstOperand_precedence == -1 || secondOperand_precedence == -1)){
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+    }
+    //check if the precedence order of the operators is swapped
+    UA_Variant *firstCompareOperand = firstOperand;
+    UA_Variant *secondCompareOperand = secondOperand;
+    if (firstOperand_precedence < secondOperand_precedence){
+        firstCompareOperand = secondOperand;
+        secondCompareOperand = firstOperand;
+    }
+    UA_SByte castRule =
+        checkTypeCastingOption(firstCompareOperand->type, secondCompareOperand->type);
+
+    if(!(castRule == 0 || castRule == 1)){
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+    }
+
+    UA_Int32 rule = -1;
```

Comment:
This is easy to misunderstand.
The castRule is -1/0/1/2. This rule here has many more possibilities.
Express the difference in the wording. And/or make it an enum. And add a comment.